README
NAME

    FFI::Platypus - Write Perl bindings to non-Perl libraries with FFI. No
    XS required.

VERSION

    version 0.42

SYNOPSIS

     use FFI::Platypus;
     
     my $ffi = FFI::Platypus->new;
     $ffi->lib(undef); # search libc
     
     # call dynamically
     $ffi->function( puts => ['string'] => 'int' )->call("hello world");
     
     # attach as a xsub and call (much faster)
     $ffi->attach( puts => ['string'] => 'int' );
     puts("hello world");

DESCRIPTION

    Platypus is a library for creating interfaces to machine code libraries
    written in languages like C, C++, Fortran, Rust, Pascal. Essentially
    anything that gets compiled into machine code. This implementation uses
    libffi to accomplish this task. libffi is battle tested by a number of
    other scripting and virtual machine languages, such as Python and Ruby
    to serve a similar role. There are a number of reasons why you might
    want to write an extension with Platypus instead of XS:

    FFI / Platypus does not require messing with the guts of Perl

      XS is less of an API and more of the guts of perl splayed out to do
      whatever you want. That may at times be very powerful, but it can
      also be a frustrating exercise in hair pulling.

    FFI / Platypus is portable

      Lots of languages have FFI interfaces, and it is subjectively easier
      to port an extension written in FFI in Perl or another language to
      FFI in another language or Perl. One goal of the Platypus Project is
      to reduce common interface specifications to a common format like
      JSON that could be shared between different languages.

    FFI / Platypus could be a bridge to Perl 6

      One of those "other" languages could be Perl 6 and Perl 6 already has
      an FFI interface I am told.

    FFI / Platypus can be reimplemented

      In a bright future with multiple implementations of Perl 5, each
      interpreter will have its own implementation of Platypus, allowing
      extensions to be written once and used on multiple platforms, in much
      the same way that Ruby-FFI extensions can be use in Ruby, JRuby and
      Rubinius.

    FFI / Platypus is pure perl (sorta)

      One Platypus script or module works on any platform where the
      libraries it uses are available. That means you can deploy your
      Platypus script in a shared filesystem where they may be run on
      different platforms. It also means that Platypus modules do not need
      to be installed in the platform specific Perl library path.

    FFI / Platypus is not C or C++ centric

      XS is implemented primarily as a bunch of C macros, which requires at
      least some understanding of C, the C pre-processor, and some C++
      caveats (since on some platforms Perl is compiled and linked with a
      C++ compiler). Platypus on the other hand could be used to call other
      compiled languages, like Fortran, Rust, Pascal, C++, or even
      assembly, allowing you to focus on your strengths.

    FFI / Platypus does not require a parser

      Inline isolates the extension developer from XS to some extent, but
      it also requires a parser. The various Inline language bindings are a
      great technical achievement, but I think writing a parser for every
      language that you want to interface with is a bit of an anti-pattern.

    This document consists of an API reference, a set of examples, some
    support and development (for contributors) information. If you are new
    to Platypus or FFI, you may want to skip down to the EXAMPLES to get a
    taste of what you can do with Platypus.

    Platypus also provides an declarative interface you may want to use
    instead of the object oriented interface called FFI::Platypus::Declare.

    Platypus has extensive documentation of types at FFI::Platypus::Type
    and its custom types API at FFI::Platypus::API.

CONSTRUCTORS

 new

     my $ffi = FFI::Platypus->new(%options);

    Create a new instance of FFI::Platypus.

    Any types defined with this instance will be valid for this instance
    only, so you do not need to worry about stepping on the toes of other
    CPAN FFI / Platypus Authors.

    Any functions found will be out of the list of libraries specified with
    the lib attribute.

  options

    lib

      Either a pathname (string) or a list of pathnames (array ref of
      strings) to pre-populate the lib attribute.

    ignore_not_found

      [version 0.15]

      Set the ignore_not_found attribute.

    lang

      [version 0.18]

      Set the lang attribute.

ATTRIBUTES

 lib

     $ffi->lib($path1, $path2, ...);
     my @paths = $ffi->lib;

    The list of libraries to search for symbols in.

    The most portable and reliable way to find dynamic libraries is by
    using FFI::CheckLib, like this:

     use FFI::CheckLib 0.06;
     $ffi->lib(find_lib_or_die lib => 'archive'); 
       # finds libarchive.so on Linux
       #       libarchive.bundle on OS X
       #       libarchive.dll (or archive.dll) on Windows
       #       cygarchive-13.dll on Cygwin
       #       ...
       # and will die if it isn't found

    FFI::CheckLib has a number of options, such as checking for specific
    symbols, etc. You should consult the documentation for that module.

    As a special case, if you add undef as a "library" to be searched,
    Platypus will also search the current process for symbols. This is
    mostly useful for finding functions in the standard C library, without
    having to know the name of the standard c library for your platform (as
    it turns out it is different just about everywhere!).

    You may also use the "find_lib" method as a shortcut:

     $ffi->find_lib( lib => 'archive' );

 ignore_not_found

    [version 0.15]

     $ffi->ignore_not_found(1);
     my $ignore_not_found = $ffi->ignore_not_found;

    Normally the attach and function methods will throw an exception if it
    cannot find the name of the function you provide it. This will change
    the behavior such that function will return undef when the function is
    not found and attach will ignore functions that are not found. This is
    useful when you are writing bindings to a library and have many
    optional functions and you do not wish to wrap every call to function
    or attach in an eval.

 lang

    [version 0.18]

     $ffi->lang($language);

    Specifies the foreign language that you will be interfacing with. The
    default is C. The foreign language specified with this attribute
    changes the default native types (for example, if you specify Rust, you
    will get i32 as an alias for sint32 instead of int as you do with C).

    If the foreign language plugin supports it, this will also enable
    Platypus to find symbols using the demangled names (for example, if you
    specify CPP for C++ you can use method names like Foo::get_bar() with
    "attach" or "function".

METHODS

 type

     $ffi->type($typename);
     $ffi->type($typename => $alias);

    Define a type. The first argument is the native or C name of the type.
    The second argument (optional) is an alias name that you can use to
    refer to this new type. See FFI::Platypus::Type for legal type
    definitions.

    Examples:

     $ffi->type('sint32'); # oly checks to see that sint32 is a valid type
     $ffi->type('sint32' => 'myint'); # creates an alias myint for sint32
     $ffi->type('bogus'); # dies with appropriate diagnostic

 custom_type

     $ffi->custom_type($alias => {
       native_type         => $native_type,
       native_to_perl      => $coderef,
       perl_to_native      => $coderef,
       perl_to_native_post => $coderef,
     });

    Define a custom type. See FFI::Platypus::Type#Custom-Types for details.

 load_custom_type

     $ffi->load_custom_type($name => $alias, @type_args);

    Load the custom type defined in the module $name, and make an alias
    $alias. If the custom type requires any arguments, they may be passed
    in as @type_args. See FFI::Platypus::Type#Custom-Types for details.

    If $name contains :: then it will be assumed to be a fully qualified
    package name. If not, then FFI::Platypus::Type:: will be prepended to
    it.

 types

     my @types = $ffi->types;
     my @types = FFI::Platypus->types;

    Returns the list of types that FFI knows about. This will include the
    native libffi types (example: sint32, opaque and double) and the normal
    C types (example: unsigned int, uint32_t), any types that you have
    defined using the type method, and custom types.

    The list of types that Platypus knows about varies somewhat from
    platform to platform, FFI::Platypus::Type includes a list of the core
    types that you can always count on having access to.

    It can also be called as a class method, in which case, no user defined
    or custom types will be included in the list.

 type_meta

     my $meta = $ffi->type_meta($type_name);
     my $meta = FFI::Platypus->type_meta($type_name);

    Returns a hash reference with the meta information for the given type.

    It can also be called as a class method, in which case, you won't be
    able to get meta data on user defined types.

    The format of the meta data is implementation dependent and subject to
    change. It may be useful for display or debugging.

    Examples:

     my $meta = $ffi->type_meta('int');        # standard int type
     my $meta = $ffi->type_meta('int[64]');    # array of 64 ints
     $ffi->type('int[128]' => 'myintarray');
     my $meta = $ffi->type_meta('myintarray'); # array of 128 ints

 function

     my $function = $ffi->function($name => \@argument_types => $return_type);
     my $function = $ffi->function($address => \@argument_types => $return_type);

    Returns an object that is similar to a code reference in that it can be
    called like one.

    Caveat: many situations require a real code reference, so at the price
    of a performance penalty you can get one like this:

     my $function = $ffi->function(...);
     my $coderef = sub { $function->(@_) };

    It may be better, and faster to create a real Perl function using the
    attach method.

    In addition to looking up a function by name you can provide the
    address of the symbol yourself:

     my $address = $ffi->find_symbol('my_functon');
     my $function = $ffi->function($address => ...);

    Under the covers, function uses find_symbol when you provide it with a
    name, but it is useful to keep this in mind as there are alternative
    ways of obtaining a functions address. Example: a C function could
    return the address of another C function that you might want to call,
    or modules such as FFI::TinyCC produce machine code at runtime that you
    can call from Platypus.

    Examples:

     my $function = $ffi->function('my_function_name', ['int', 'string'] => 'string');
     my $return_string = $function->(1, "hi there");

 attach

     $ffi->attach($name => \@argument_types => $return_type);
     $ffi->attach([$c_name => $perl_name] => \@argument_types => $return_type);
     $ffi->attach([$address => $perl_name] => \@argument_types => $return_type);
     $ffi->attach($name => \@argument_types => $return_type, sub { ... });
     $ffi->attach([$c_name => $perl_name] => \@argument_types => $return_type, sub { ... });
     $ffi->attach([$address => $perl_name] => \@argument_types => $return_type, sub { ... });

    Find and attach a C function as a real live Perl xsub. The advantage of
    attaching a function over using the function method is that it is much
    much much faster since no object resolution needs to be done. The
    disadvantage is that it locks the function and the FFI::Platypus
    instance into memory permanently, since there is no way to deallocate
    an xsub.

    If just one $name is given, then the function will be attached in Perl
    with the same name as it has in C. The second form allows you to give
    the Perl function a different name. You can also provide an address
    (the third form), just like with the function method.

    Examples:

     $ffi->attach('my_functon_name', ['int', 'string'] => 'string');
     $ffi->attach(['my_c_functon_name' => 'my_perl_function_name'], ['int', 'string'] => 'string');
     my $string1 = my_function_name($int);
     my $string2 = my_perl_function_name($int);

    [version 0.20]

    If the last argument is a code reference, then it will be used as a
    wrapper around the attached xsub. The first argument to the wrapper
    will be the inner xsub. This can be used if you need to verify/modify
    input/output data.

    Examples:

     $ffi->attach('my_function', ['int', 'string'] => 'string', sub {
       my($my_function_xsub, $integer, $string) = @_;
       $integer++;
       $string .= " and another thing";
       my $return_string = $my_function_xsub->($integer, $string);
       $return_string =~ s/Belgium//; # HHGG remove profanity
       $return_string;
     });

 closure

     my $closure = $ffi->closure($coderef);

    Prepares a code reference so that it can be used as a FFI closure (a
    Perl subroutine that can be called from C code). For details on
    closures, see FFI::Platypus::Type#Closures.

 cast

     my $converted_value = $ffi->cast($original_type, $converted_type, $original_value);

    The cast function converts an existing $original_value of type
    $original_type into one of type $converted_type. Not all types are
    supported, so care must be taken. For example, to get the address of a
    string, you can do this:

     my $address = $ffi->cast('string' => 'opaque', $string_value);

    Something that won't work is trying to cast an array to anything:

     my $address = $ffi->cast('int[10]' => 'opaque', \@list);  # WRONG

 attach_cast

     $ffi->attach_cast("cast_name", $original_type, $converted_type);
     my $converted_value = cast_name($original_value);

    This function attaches a cast as a permanent xsub. This will make it
    faster and may be useful if you are calling a particular cast a lot.

 sizeof

     my $size = $ffi->sizeof($type);

    Returns the total size of the given type in bytes. For example to get
    the size of an integer:

     my $intsize = $ffi->sizeof('int');   # usually 4
     my $longsize = $ffi->sizeof('long'); # usually 4 or 8 depending on platform

    You can also get the size of arrays

     my $intarraysize = $ffi->sizeof('int[64]');  # usually 4*64
     my $intarraysize = $ffi->sizeof('long[64]'); # usually 4*64 or 8*64
                                                  # depending on platform

    Keep in mind that "pointer" types will always be the pointer / word
    size for the platform that you are using. This includes strings, opaque
    and pointers to other types.

    This function is not very fast, so you might want to save this value as
    a constant, particularly if you need the size in a loop with many
    iterations.

 alignof

    [version 0.21]

     my $align = $ffi->alignof($type);

    Returns the alignment of the given type in bytes.

 find_lib

    [version 0.20]

     $ffi->find_lib( lib => $libname );

    This is just a shortcut for calling FFI::CheckLib#find_lib and updating
    the "lib" attribute appropriately. Care should be taken though, as this
    method simply passes its arguments to FFI::CheckLib#find_lib, so if
    your module or script is depending on a specific feature in
    FFI::CheckLib then make sure that you update your prerequisites
    appropriately.

 find_symbol

     my $address = $ffi->find_symbol($name);

    Return the address of the given symbol (usually function).

 package

    [version 0.15]

     $ffi->package($package, $file); # usually __PACKAGE__ and __FILE__ can be used
     $ffi->package;                  # autodetect

    If you have used Module::Build::FFI to bundle C code with your
    distribution, you can use this method to tell the FFI::Platypus
    instance to look for symbols that came with the dynamic library that
    was built when your distribution was installed.

 abis

     my $href = $ffi->abis;
     my $href = FFI::Platypus->abis;

    Get the legal ABIs supported by your platform and underlying
    implementation. What is supported can vary a lot by CPU and by
    platform, or even between 32 and 64 bit on the same CPU and platform.
    They keys are the "ABI" names, also known as "calling conventions". The
    values are integers used internally by the implementation to represent
    those ABIs.

 abi

     $ffi->abi($name);

    Set the ABI or calling convention for use in subsequent calls to
    "function" or "attach". May be either a string name or integer value
    from the "abis" method above.

EXAMPLES

    Here are some examples. Some of them use the FFI::Platypus::Declare
    interface, but the principles apply to the OO interface. These examples
    are provided in full with the Platypus distribution in the "examples"
    directory. There are also some more examples in FFI::Platypus::Type
    that are related to types.

 Integer conversions

     use FFI::Platypus::Declare qw( int string );
     
     lib undef;
     attach puts => [string] => int;
     attach atoi => [string] => int;
     
     puts(atoi('56'));

    Discussion: puts and atoi should be part of the standard C library on
    all platforms. puts prints a string to standard output, and atoi
    converts a string to integer. Specifying undef as a library tells
    Platypus to search the current process for symbols, which includes the
    standard c library.

 libnotify

     use FFI::CheckLib;
     use FFI::Platypus::Declare qw( void string opaque );
     
     # NOTE: I ported this from the like named eg/notify.pl that came with FFI::Raw
     # and it seems to work most of the time, but also seems to SIGSEGV sometimes.
     # I saw the same behavior in the FFI::Raw version, and am not really familiar
     # with the libnotify API to say what is the cause.  Patches welcome to fix it.
     
     lib find_lib_or_exit lib => 'notify';
     
     attach notify_init   => [string] => void;
     attach notify_uninit => []       => void;
     attach [notify_notification_new    => 'notify_new']    => [string,string,string]           => opaque;
     attach [notify_notification_update => 'notify_update'] => [opaque, string, string, string] => void;
     attach [notify_notification_show   => 'notify_show']   => [opaque, opaque]                 => void;
     
     notify_init('FFI::Platypus');
     my $n = notify_new('','','');
     notify_update($n, 'FFI::Platypus', 'It works!!!', 'media-playback-start');
     notify_show($n, undef);
     notify_uninit();

    Discussion: libnotify is a desktop GUI notification library for the
    GNOME Desktop environment. This script sends a notification event that
    should show up as a balloon, for me it did so in the upper right hand
    corner of my screen.

    The most portable way to find the correct name and location of a
    dynamic library is via the FFI::CheckLib#find_lib family of functions.
    If you are putting together a CPAN distribution, you should also
    consider using FFI::CheckLib#check_lib_or_exit function in your
    Build.PL or Makefile.PL file (If you are using Dist::Zilla, check out
    the Dist::Zilla::Plugin::FFI::CheckLib plugin). This will provide a
    user friendly diagnostic letting the user know that the required
    library is missing, and reduce the number of bogus CPAN testers results
    that you will get.

    Also in this example, we rename some of the functions when they are
    placed into Perl space to save typing:

     attach [notify_notification_new => 'notify_new']
       => [string,string,string]
       => opaque;

    When you specify a list reference as the "name" of the function the
    first element is the symbol name as understood by the dynamic library.
    The second element is the name as it will be placed in Perl space.

    Later, when we call notify_new:

     my $n = notify_new('','','');

    We are really calling the C function notify_notification_new.

 Allocating and freeing memory

     use FFI::Platypus::Declare;
     use FFI::Platypus::Memory qw( malloc free memcpy );
     
     my $buffer = malloc 12;
     
     memcpy $buffer, cast('string' => 'opaque', "hello there"), length "hello there\0";
     
     print cast('opaque' => 'string', $buffer), "\n";
     
     free $buffer;

    Discussion: malloc and free are standard memory allocation functions
    available from the standard c library and. Interfaces to these and
    other memory related functions are provided by the
    FFI::Platypus::Memory module.

 structured data records

     package My::UnixTime;
     
     use FFI::Platypus::Record;
     
     record_layout(qw(
         int    tm_sec
         int    tm_min
         int    tm_hour
         int    tm_mday
         int    tm_mon
         int    tm_year
         int    tm_wday
         int    tm_yday
         int    tm_isdst
         long   tm_gmtoff
         string tm_zone
     ));
     
     my $ffi = FFI::Platypus->new;
     $ffi->lib(undef);
     # define a record class My::UnixTime and alias it to "tm"
     $ffi->type("record(My::UnixTime)" => 'tm');
     
     # attach the C localtime function as a constructor
     $ffi->attach( localtime => ['time_t*'] => 'tm', sub {
       my($inner, $class, $time) = @_;
       $time = time unless defined $time;
       $inner->(\$time);
     });
     
     package main;
     
     # now we can actually use our My::UnixTime class
     my $time = My::UnixTime->localtime;
     printf "time is %d:%d:%d %s\n",
       $time->tm_hour,
       $time->tm_min,
       $time->tm_sec,
       $time->tm_zone;

    Discussion: C and other machine code languages frequently provide
    interfaces that include structured data records (known as "structs" in
    C). They sometimes provide an API in which you are expected to
    manipulate these records before and/or after passing them along to C
    functions. There are a few ways of dealing with such interfaces, but
    the easiest way is demonstrated here defines a record class using a
    specific layout. For more details see FFI::Platypus::Record.
    (FFI::Platypus::Type includes some other ways of manipulating
    structured data records).

 libuuid

     use FFI::CheckLib;
     use FFI::Platypus::Declare
       'void',
       [ 'string(37)' => 'uuid_string' ],
       [ 'record(16)' => 'uuid_t' ];
     use FFI::Platypus::Memory qw( malloc free );
     
     lib find_lib_or_exit lib => 'uuid';
     
     attach uuid_generate => [uuid_t] => void;
     attach uuid_unparse  => [uuid_t,uuid_string] => void;
     
     my $uuid = "\0" x 16;  # uuid_t
     uuid_generate($uuid);
     
     my $string = "\0" x 37; # 36 bytes to store a UUID string 
                             # + NUL termination
     uuid_unparse($uuid, $string);
     
     print "$string\n";

    Discussion: libuuid is a library used to generate unique identifiers
    (UUID) for objects that may be accessible beyond the local system. The
    library is or was part of the Linux e2fsprogs package.

    Knowing the size of objects is sometimes important. In this example, we
    use the sizeof function to get the size of 16 characters (in this case
    it is simply 16 bytes). We also know that the strings "deparsed" by
    uuid_unparse are exactly 37 bytes.

 puts and getpid

     use FFI::Platypus::Declare qw( string int );
     
     lib undef;
     attach puts => [string] => int;
     attach getpid => [] => int;
     
     puts(getpid());

    Discussion: puts is part of standard C library on all platforms. getpid
    is available on Unix type platforms.

 Math library

     use FFI::Platypus::Declare qw( string int double );
     use FFI::CheckLib;
     
     lib undef;
     attach puts => [string] => int;
     attach fdim => [double,double] => double;
     
     puts(fdim(7.0, 2.0));
     
     attach cos => [double] => double;
     
     puts(cos(2.0));
     
     attach fmax => [double, double] => double;
     
     puts(fmax(2.0,3.0));

    Discussion: On UNIX the standard c library math functions are
    frequently provided in a separate library libm, so you could search for
    those symbols in "libm.so", but that won't work on non-UNIX platforms
    like Microsoft Windows. Fortunately Perl uses the math library so these
    symbols are already in the current process so you can use undef as the
    library to find them.

 Strings

     use FFI::Platypus::Declare qw( int string );
     
     lib undef;
     attach puts => [string] => int;
     attach strlen => [string] => int;
     
     puts(strlen('somestring'));
     
     attach strstr => [string,string] => string;
     
     puts(strstr('somestring', 'string'));
     
     #attach puts => [string] => int;
     
     puts(puts("lol"));
     
     attach strerror => [int] => string;
     
     puts(strerror(2));

    Discussion: Strings are not a native type to libffi but the are handled
    seamlessly by Platypus.

 Attach function from pointer

     use FFI::TinyCC;
     use FFI::Platypus;
     
     my $ffi = FFI::Platypus->new;
     my $tcc = FFI::TinyCC->new;
     
     $tcc->compile_string(q{
       int
       add(int a, int b)
       {
         return a+b;
       }
     });
     
     my $address = $tcc->get_symbol('add');
     
     $ffi->attach( [ $address => 'add' ] => ['int','int'] => 'int' );
     
     print add(1,2), "\n";

    Discussion: Sometimes you will have a pointer to a function from a
    source other than Platypus that you want to call. You can use that
    address instead of a function name for either of the function or attach
    methods. In this example we use FFI::TinyCC to compile a short piece of
    C code and to give us the address of one of its functions, which we
    then use to create a perl xsub to call it.

    FFI::TinyCC embeds the Tiny C Compiler (tcc) to provide a just-in-time
    (JIT) compilation service for FFI.

 libzmq

     use constant ZMQ_IO_THREADS  => 1;
     use constant ZMQ_MAX_SOCKETS => 2;
     use constant ZMQ_REQ => 3;
     use constant ZMQ_REP => 4;
     use FFI::CheckLib qw( find_lib_or_exit );
     use FFI::Platypus;
     use FFI::Platypus::Memory qw( malloc );
     use FFI::Platypus::Buffer qw( scalar_to_buffer buffer_to_scalar );
     
     my $endpoint = "ipc://zmq-ffi-$$";
     my $ffi = FFI::Platypus->new;
     
     $ffi->lib(undef); # for puts
     $ffi->attach(puts => ['string'] => 'int');
     
     $ffi->lib(find_lib_or_exit lib => 'zmq');
     $ffi->attach(zmq_version => ['int*', 'int*', 'int*'] => 'void');
     
     my($major,$minor,$patch);
     zmq_version(\$major, \$minor, \$patch);
     puts("libzmq version $major.$minor.$patch");
     die "this script only works with libzmq 3 or better" unless $major >= 3;
     
     $ffi->type('opaque'       => 'zmq_context');
     $ffi->type('opaque'       => 'zmq_socket');
     $ffi->type('opaque'       => 'zmq_msg_t');
     $ffi->attach(zmq_ctx_new  => [] => 'zmq_context');
     $ffi->attach(zmq_ctx_set  => ['zmq_context', 'int', 'int'] => 'int');
     $ffi->attach(zmq_socket   => ['zmq_context', 'int'] => 'zmq_socket');
     $ffi->attach(zmq_connect  => ['opaque', 'string'] => 'int');
     $ffi->attach(zmq_bind     => ['zmq_socket', 'string'] => 'int');
     $ffi->attach(zmq_send     => ['zmq_socket', 'opaque', 'size_t', 'int'] => 'int');
     $ffi->attach(zmq_msg_init => ['zmq_msg_t'] => 'int');
     $ffi->attach(zmq_msg_recv => ['zmq_msg_t', 'zmq_socket', 'int'] => 'int');
     $ffi->attach(zmq_msg_data => ['zmq_msg_t'] => 'opaque');
     $ffi->attach(zmq_errno    => [] => 'int');
     $ffi->attach(zmq_strerror => ['int'] => 'string');
     
     my $context = zmq_ctx_new();
     zmq_ctx_set($context, ZMQ_IO_THREADS, 1);
     
     my $socket1 = zmq_socket($context, ZMQ_REQ);
     zmq_connect($socket1, $endpoint);
     
     my $socket2 = zmq_socket($context, ZMQ_REP);
     zmq_bind($socket2, $endpoint);
     
     do { # send
       our $sent_message = "hello there";
       my($pointer, $size) = scalar_to_buffer $sent_message;
       my $r = zmq_send($socket1, $pointer, $size, 0);
       die zmq_strerror(zmq_errno()) if $r == -1;
     };
     

Changes
Revision history for FFI-Platypus

0.42      2016-05-06 16:31:01 -0400
  - Support for MSYS2

0.41      2016-04-09 16:03:07 -0400
  - For the return value undef can be passed in to mean 'void'.
  - Fixed installer bug where ABI probe would silently fail if /tmp
    was mounted noexec
  - Avoid unnecessary downgrade on when Perl is compiled to use longdouble

0.40      2015-08-29 08:45:19 -0400
  - Closure declerations ignore white space between () and ->

0.39      2015-08-24 03:23:10 -0400
  - Fixed spurious warning: auto loading of record class (via FFI::Platypus::Record)
    was ALWAYS warning, when it should have only been warning on load failure.

0.38      2015-08-13 17:13:07 -0400
  - closure method now comes with a more useful diagnostic thinks to
    Carp::croak
  - Added a check for standard C headers.  This seems to help the configure
    stage find ptrdiff_t, which was frequently not detected correctly
    (at least on Linux).
  - Improved thread safety by using MY_CXT for some very infrequently used
    global variables.
  - Added IRC meta data information for metacpan.org.  Please join us at
    #native on irc.perl.org!
  - Many minor documentation corrections and tweaks.  Most significant is
    that Convert::Binary::C can now be recommended as it is once again
    properly maintained.
  - Added tests for threads and forks.  If these tests fail in your environment
    please let me know!

0.37      2015-05-29 14:28:21 -0400
  - Added compatability back in for older version of constant
    (newer one is not available on CPAN yet)

0.36      2015-05-29 13:40:32 -0400
  - Explicitly require constant pragma version 1.32

0.35      2015-05-29 12:06:39 -0400
  - FFI::Platypus::Record uses constant to create size of alignment constants
    instead of creating them with a sub reference (this usage was deprecated
    in Perl 5.22)

0.34      2015-05-07 09:27:04 -0400
  - Require Alien::FFI which is more reliable at configure time on some platforms

0.33      2015-03-23 21:55:02 -0400
  - Additional fix for Microsoft Visual C++ that didn't get folded into
    the previous release.
  - Fixed segfault during global destruction (gh#53)

0.32      2015-03-18 13:02:53 -0400
  - Make sure -L flags from Alien::FFI come before those in perl Config
    For more reliable builds
  - Support for Microsoft Visual C++ (you will probably also need
    Alien::FFI 0.09 or beter)

0.31      2015-02-26 13:41:23 -0500
  - Fix bug involving wide custom arguments 
    "wide" meaning where a single Perl argument is translated
    into multiple machine code arguments.  (pipcet++ gh#43)

0.30      2015-02-25 17:50:54 -0500
  - You can now pass an opaque in place of a closure type (pipcet++ gh#40,gh#41)
  - FFI closures are now cached and can be reused if the same closure is passed
    repeatedly (pipcet++ gh#40,gh#42)
  - Passing non-reference to scalar as a pointer argument will now issue a warning
    (gh#5)

0.29      2015-02-24 08:50:34 -0500
  - Delayed loading of Win32::ErrorMode to avoid build prereq failure on Windows

0.28      2015-02-23 14:01:54 -0500
  - Fix Win32 probe prereq on non-Strawberry 5.20.x+
  - Fix for Solaris cc

0.27      2015-02-22 11:17:05 -0500
  - Interface to alternate ABIs / calling conventions
  - Added abi method
  - Added abis class method
  - Simplify Win32 probe
  - Added FFI::Platypus::Lang::Win32 which provides data types used by
    the Windows API.  Takes care of subtle differences between
    Win32 and Win64.
  - Fixed bugs specific to 5.8.x
  - Language plugins can now specify an ABI with abi class method
  - Default ABI for FFI::Platypus::Lang::Win32 is stdcall on 32bit windows

0.26      2015-02-18 17:47:43 -0500
  - Added support for pointers to longdouble (in C "long double *") type
  - Added support for array of longdouble (in C "long double []") type
  - Added tied array interface for record array members (see FFI::Platypus::Record::TieArray)
    Marked as EXPERIMENTAL
  - Array members of records can now be accessed (set/get) by element
  - Array members of records types are now documented (see FFI::Platypus::Record)
  - Bugfix: array wasn't being updated on return for variable length
    array types
  - Should now build with an Alien::FFI that was built with ALIEN_FORCE=1

0.25      2015-02-16 20:18:41 -0500
  - Probe for proper long double support instead of trusting ffi.h
  - This disables long double support on cygwin, which does not seem to
    work, at least in so far as it seems to work on other platforms
    patches to prove otherwise are welcome.

0.24      2015-02-16 15:38:58 -0500
  - Fixed Windows / Strawberry configuration issues

0.23      2015-02-16 05:44:39 -0500
  - Support for longdouble (in C "long double") type.
  - Support for complex_float (in C "float complex") type
  - Support for complex_double (in C "double complex") type
  - Fixes for Big Endian architectures (tested on Linux PowerPC)

0.22      2015-02-12 07:47:32 -0500
  - Variable length arrays
  - More recent version of Config::AutoConf required in the configure step
    (gh#33 zmughal)
  - Documentation improvements and additional examples, including a 
    crazy Java one

0.21      2015-02-09 06:23:03 -0500
  - Added FFI::Platypus#alignof method
  - Added FFI::Platypus::Record module
  - Added fixed length strings example: string(10)
  - Added ro and rw trait for strings

0.20      2015-02-05 14:06:11 -0500
  - Added optional wrapper argument to FFI::Platypus#attach
    and FFI::Platypus::Declare#attach
  - Added FFI::Platypus#find_lib method
  - FFI::CheckLib is now a runtime requirement for Platypus
  - Bumped Alien::FFI requirement to 0.06

0.19      2015-02-03 13:34:53 -0500
  - Accept additional extensions, in addition to dlext
    Example: on OS X both .bundle and .dylib can be used
    Example: although arguably wrong, on cygwin sometimes .so is used
  - Added Module::Build::FFI->ffi_dlext class method

0.18      2015-01-30 15:22:07 -0500

Build.PL
use strict; use warnings;
BEGIN {
  unless(eval q{ use 5.008001; 1}) {
    print "Perl 5.008001 or better required\n";
    exit;
  }
}

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.003.
use strict;
use warnings;

use Module::Build 0.3601;
use lib qw{inc}; use My::ModuleBuild;

my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Alien::FFI" => "0.12",
    "Config::AutoConf" => "0.309",
    "ExtUtils::CBuilder" => 0,
    "FFI::CheckLib" => "0.05",
    "Module::Build" => "0.3601",
    "perl" => "5.008001"
  },
  "dist_abstract" => "Write Perl bindings to non-Perl libraries with FFI. No XS required.",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "FFI-Platypus",
  "dist_version" => "0.42",
  "license" => "perl",
  "module_name" => "FFI::Platypus",
  "recursive_test_files" => 1,
  "requires" => {
    "ExtUtils::CBuilder" => 0,
    "FFI::CheckLib" => 0,
    "File::ShareDir" => 0,
    "JSON::PP" => 0,
    "Module::Build" => "0.3601",
    "constant" => "1.32",
    "perl" => "5.008001"
  },
  "share_dir" => {
    "dist" => "share"
  },
  "test_requires" => {
    "Alien::FFI" => "0.012",
    "Test::More" => "0.94",
    "perl" => "5.008001"
  }
);


my %fallback_build_requires = (
  "Alien::FFI" => "0.012",
  "Module::Build" => "0.3601",
  "Test::More" => "0.94",
  "perl" => "5.008001"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = My::ModuleBuild->new(%module_build_args);


$build->create_build_script;
